layout computation: gracefully handle unsized types in unexpected locations

lukas-code opened the pull request:
This PR reworks the layout computation to eagerly return an error when encountering an unsized field where a sized field was expected, rather than delaying a bug and attempting to recover a layout. This is required, because with trivially false where clauses like `[T]: Sized`, any field can possible be an unsized type, without causing a compile error.

Since this PR removes the `delayed_bug` method from the `LayoutCalculator` trait, it essentially becomes the same as the `HasDataLayout` trait, so I've also refactored the `LayoutCalculator` to be a simple wrapper struct around a type that implements `HasDataLayout`.

The majority of the diff is whitespace changes, so viewing with whitespace ignored is advised.

implements https://github.com/rust-lang/rust/pull/123169#issuecomment-2025788480

r? @compiler-errors or compiler

fixes https://github.com/rust-lang/rust/issues/123134
fixes https://github.com/rust-lang/rust/issues/124182
fixes https://github.com/rust-lang/rust/issues/126939
fixes https://github.com/rust-lang/rust/issues/127737

Veykril reviewed (approved):
r-a side

Commit:
make `LayoutCx` not generic

Commit:
layout computation: eagerly error for unexpected unsized fields

bors commented:
:umbrella: The latest upstream changes (presumably #130415) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

compiler-errors commented on compiler/rustc_ty_utils/src/layout.rs:
```diff
@@ -80,7 +80,36 @@ fn layout_of<'tcx>(
+            LayoutError::SizeOverflow(ty)
+        }
+        LayoutCalculatorError::UnexpectedUnsized => {
+            // This is sometimes not a compile error if there are trivially false where
+            // clauses, but it is always a compiler error in the empty environment.
+            if cx.param_env.caller_bounds().is_empty() {
+                cx.tcx().dcx().delayed_bug(format!(
+                    "encountered unexpected unsized field in layout of {ty:?}"
+                ));
+            }
```
We could perhaps attempt to prove that the type is sized here.

compiler-errors replied:
Also would be useful to point to an example.

lukas-code replied:
> We could perhaps attempt to prove that the type is sized here.

The final type is not always guaranteed to be `Sized` (in the type system way), for example given
```rust
#![feature(trivial_bounds)]
struct Foo
where
    [u8]: Sized,
{
    a: [u8],
    b: [u16],
}
```

the trait bound `Foo: Sized` will not hold, but attempting to compute the layout of `Foo` will error with `UnexpectedUnsized`.

compiler-errors replied:
Right, but shouldn't we be able to prove that the type is sized within the param-env we're computing the layout? Or otherwise, why would we care that the param-env has no clauses?

compiler-errors replied:
Like, I'd want to double check that the param-env in fact does contain `[u8]: Sized` rather than just checking that the param-env is non-empty.

lukas-code replied:
Right, makes sense to check for the field itself. I'll try to add this check.

lukas-code replied:
Ok, I've replaced the check for an empty param-env with a check that the layout-unsized field is type-system-sized now.

Also, looking at the code again, I found that an old hack could now be replaced, so I did that in another commit.

compiler-errors commented on compiler/rustc_ty_utils/src/layout.rs:
```diff
@@ -80,7 +80,36 @@ fn layout_of<'tcx>(
+) -> &'tcx LayoutError<'tcx> {
+    let err = match err {
+        LayoutCalculatorError::SizeOverflow => {
+            // This is sometimes not a compile error in `check` builds.
```
Yeah, it wouldn't hurt.

Commit:
assert that unexpectedly unsized fields are sized in the param env

Commit:
get rid of an old hack

For structs that cannot be unsized, the layout algorithm sometimes moves
unsized fields to the end of the struct, which circumvented the error
for unexpected unsized fields and returned an unsized layout anyway.

This commit makes it so that the unexpected unsized error is always
returned for structs that cannot be unsized, allowing us to remove an
old hack and fixing some old ICE.

compiler-errors commented on compiler/rustc_abi/src/layout.rs:
```diff
@@ -97,16 +99,17 @@ impl<Cx: HasDataLayout> LayoutCalculator<Cx> {
 
     pub fn univariant<
-        'a,
+        'fields,
```
Huh... is it possible for us to just require `F: Copy` rather than passing around `&'fields F` everywhere? It seems a bit redundant.

lukas-code replied:
Yes that's possible and much cleaner. Idk why I tried so hard to make lifetimes happen here :upside_down_face:

compiler-errors commented on compiler/rustc_index/src/slice.rs:
```diff
@@ -20,7 +20,7 @@ pub struct IndexSlice<I: Idx, T> {
 impl<I: Idx, T> IndexSlice<I, T> {
     #[inline]
-    pub const fn empty() -> &'static Self {
+    pub const fn empty<'a>() -> &'a Self {
```
Is this because of invariance or what?

lukas-code replied:
Yeah this function is used here https://github.com/rust-lang/rust/blob/4edbd9320b39592a2af26ca5cb68018fadebc3fa/compiler/rustc_ty_utils/src/layout.rs#L374-L380

And changing the lifetime from `'_` to `'tcx` here https://github.com/rust-lang/rust/blob/4edbd9320b39592a2af26ca5cb68018fadebc3fa/compiler/rustc_ty_utils/src/layout.rs#L119

means that we would now create a `&'static &'tcx Something` which would require `'tcx: 'static`,

compiler-errors replied:
Ah, ok. Yeah, that makes sense 👍

compiler-errors commented on compiler/rustc_ty_utils/src/layout.rs:
```diff
@@ -79,31 +78,58 @@ fn layout_of<'tcx>(
+            // This is sometimes not a compile error if there are trivially false where clauses.
+            // See `tests/ui/layout/trivial-bounds-sized.rs` for an example.
+            assert!(field.layout.is_unsized(), "invalid layout error {err:#?}");
+            if !field.ty.is_sized(cx.tcx(), cx.param_env) {
```
👍

compiler-errors reviewed (approved):
Great work!

compiler-errors commented:
r=me when green

compiler-errors commented:
@bors r+ rollup=never

bors commented:
:pushpin: Commit 20d241492590a83eb0946ce4414255e25d2112b8 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 20d241492590a83eb0946ce4414255e25d2112b8 -->
<!-- homu: {"type":"Approved","sha":"20d241492590a83eb0946ce4414255e25d2112b8","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 20d241492590a83eb0946ce4414255e25d2112b8 with merge e2dc1a1c0f97a90319181a721ab317210307617a...
<!-- homu: {"type":"BuildStarted","head_sha":"20d241492590a83eb0946ce4414255e25d2112b8","merge_sha":"e2dc1a1c0f97a90319181a721ab317210307617a"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10894791048/job/30235403439)
Approved by: compiler-errors
Pushing e2dc1a1c0f97a90319181a721ab317210307617a to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"compiler-errors","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/10894791048/job/30235403439"},"merge_sha":"e2dc1a1c0f97a90319181a721ab317210307617a"} -->

rust-timer commented:
Finished benchmarking commit (e2dc1a1c0f97a90319181a721ab317210307617a): [comparison URL](https://perf.rust-lang.org/compare.html?start=bde6bf2b07b61f7da92c211d2b4a897ea4620e1b&end=e2dc1a1c0f97a90319181a721ab317210307617a&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=bde6bf2b07b61f7da92c211d2b4a897ea4620e1b&end=e2dc1a1c0f97a90319181a721ab317210307617a&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=bde6bf2b07b61f7da92c211d2b4a897ea4620e1b&end=e2dc1a1c0f97a90319181a721ab317210307617a&stat=max-rss)
<details>
<summary>Results (primary -2.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.3% | [-2.3%, -2.3%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -2.3% | [-2.3%, -2.3%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=bde6bf2b07b61f7da92c211d2b4a897ea4620e1b&end=e2dc1a1c0f97a90319181a721ab317210307617a&stat=cycles:u)
<details>
<summary>Results (primary -0.6%, secondary 5.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 5.2%  | [5.2%, 5.2%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.6% | [-0.6%, -0.6%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.6% | [-0.6%, -0.6%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=bde6bf2b07b61f7da92c211d2b4a897ea4620e1b&end=e2dc1a1c0f97a90319181a721ab317210307617a&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 760.616s -> 760.716s (0.01%)
**Artifact size**: 341.32 MiB -> 341.27 MiB (-0.02%)
